src/SpeedLimiter.cc

```diff
@@ -0,0 +1,201 @@
+  public: double maxJerk{std::numeric_limits<double>::infinity()};
+};
+
+//////////////////////////////////////////////////
+SpeedLimiter::SpeedLimiter(bool   _hasVelocityLimits,
+                           bool   _hasAccelerationLimits,
+                           bool   _hasJerkLimits,
+                           double _minVelocity,
+                           double _maxVelocity,
+                           double _minAcceleration,
+                           double _maxAcceleration,
+                           double _minJerk,
+                           double _maxJerk)
+  : dataPtr(std::make_unique<SpeedLimiterPrivate>(_hasVelocityLimits,
+      _hasAccelerationLimits, _hasJerkLimits, _minVelocity, _maxVelocity,
+      _minAcceleration, _maxAcceleration, _minJerk, _maxJerk))
+{
+}
+
+//////////////////////////////////////////////////
+SpeedLimiter::~SpeedLimiter() = default;
+
+//////////////////////////////////////////////////
+double SpeedLimiter::Limit(double &_v, double _v0, double _v1,
+    std::chrono::steady_clock::duration _dt) const
+{
+  const double tmp = _v;
+
+  this->LimitJerk(_v, _v0, _v1, _dt);
+  this->LimitAcceleration(_v, _v0, _dt);
+  this->LimitVelocity(_v);
+
+  if (ignition::math::equal(tmp, 0.0))
+    return 1.0;
+  else
+    return _v / tmp;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::LimitVelocity(double &_v) const
+{
+  const double tmp = _v;
+
+  if (this->dataPtr->hasVelocityLimits)
+  {
+    _v = ignition::math::clamp(
+      _v, this->dataPtr->minVelocity, this->dataPtr->maxVelocity);
+  }
+
+  if (ignition::math::equal(tmp, 0.0))
+    return 1.0;
+  else
+    return _v / tmp;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::LimitAcceleration(double &_v, double _v0,
+    std::chrono::steady_clock::duration _dt) const
+{
+  const double tmp = _v;
+
+  if (this->dataPtr->hasAccelerationLimits)
+  {
+    const double dvMin = this->dataPtr->minAcceleration *
+        std::chrono::duration<double>(_dt).count();
+    const double dvMax = this->dataPtr->maxAcceleration *
+        std::chrono::duration<double>(_dt).count();
+
+    const double dv = ignition::math::clamp(_v - _v0, dvMin, dvMax);
+
+    _v = _v0 + dv;
+  }
+
+  if (ignition::math::equal(tmp, 0.0))
+    return 1.0;
+  else
+    return _v / tmp;
+}
+
+//////////////////////////////////////////////////
+double SpeedLimiter::LimitJerk(double &_v, double _v0, double _v1,
+    std::chrono::steady_clock::duration _dt) const
+{
+  const double tmp = _v;
+
+  if (this->dataPtr->hasJerkLimits)
+  {
+    const double dv  = _v  - _v0;
+    const double dv0 = _v0 - _v1;
+
+    const double dt2 = std::chrono::duration<double>(_dt).count() *
+        std::chrono::duration<double>(_dt).count();
+
+    const double daMin = this->dataPtr->minJerk * dt2;
+    const double daMax = this->dataPtr->maxJerk * dt2;
+
+    const double da = ignition::math::clamp(dv - dv0, daMin, daMax);
```

Comment:
I think `da` could be renamed `dvFromJerk` (and `dvFromJerkMin` and `dvFromJerkMax` likewise)

Reply:
I think the new implementation looks clearer and we don't need this specific variable anymore.